buy_together_app/tests/test_login.py

```diff
@@ -0,0 +1,75 @@
+import pytest
+from unittest.mock import Mock
+from buy_together_app import views
+
+
+class Test_Login:
+    @pytest.fixture
+    def authenticate_mock(self, monkeypatch):
+        mock = Mock()
+        monkeypatch.setattr(views, views.authenticate.__name__, mock)
+        return mock
+
+    @pytest.fixture
+    def login_mock(self, monkeypatch):
+        mock = Mock()
+        monkeypatch.setattr(views, views.login.__name__, mock)
+        return mock
+
+    @pytest.fixture
+    def redirect_mock(self, monkeypatch):
+        mock = Mock()
+        monkeypatch.setattr(views, views.redirect.__name__, mock)
+        return mock
+
+    @pytest.fixture
+    def messages_mock(self, monkeypatch):
+        mock = Mock()
+        monkeypatch.setattr(views, "messages", mock)
+        return mock
+
+    @pytest.fixture
+    def render_mock(self, monkeypatch):
+        mock = Mock()
+        monkeypatch.setattr(views, views.render.__name__, mock)
+        return mock
+
+    def test_login_success(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "POST"
+        output = views.log_in_page(request_mock)
+        assert output == redirect_mock.return_value
+        assert authenticate_mock.call_count == 1
+        assert login_mock.call_count == 1
+        assert redirect_mock.call_count == 1
+        assert messages_mock.info.call_count == 0
+        assert render_mock.call_count == 0
+
+    def test_login_not_post(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "NOT_POST"
+        output = views.log_in_page(request_mock)
+        assert output == render_mock.return_value
+        assert authenticate_mock.call_count == 0
+        assert login_mock.call_count == 0
+        assert redirect_mock.call_count == 0
+        assert messages_mock.info.call_count == 0
+        assert render_mock.call_count == 1
+
+    def test_login_failed_auth(
+        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
+    ):
+        request_mock = Mock()
+        request_mock.method = "POST"
+        authenticate_mock.return_value = None
+        output = views.log_in_page(request_mock)
+        assert output == render_mock.return_value
+        assert authenticate_mock.call_count == 1
```

Comment:
you also have `assert_called_once` or `assert_called_once_with` for that

Reply:
@pazMenachem just a word of caution. To keep things simple, the code I suggested was creating mocks directly with `Mock()`. A better approach, as Omer mentioned, is to use `create_autospec` which takes an object as specification and creates a mock containing only the functions the spec has (so calling functions which doesn't exist will fail - as with regular objects).

When not autospecing, your mock will allow calling any function, and asserting on that call will always pass.

So, suppose that you want to use `mock.assert_called_once()` if you happen to misspell it, then your test will always pass regardless of whether the function was called or not.

Bottom line, if you want to use the mock's functions, e.g. `assert_called_once()` (very useful), then only do it when the mock has a spec, and not in the case of creating a blank mock like we did here.

Reply:
im sure there are better ways, but since we are low on time the code just suppose to do the trick.
since i haven't done any testing of this kind i will learn how to do so in the weekend, but for now i just wanted this part to be merged.